### PR TITLE
add ddp sync for logging in result step

### PR DESCRIPTION
**Pull request by justusschock**
## What does this PR do?

Adds the possibility to sync between DDP Processes with the result object

# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by justusschock** [comment]
I will fix pep8 and retriever tests once #2568 is merged :)

**Comment by Borda** [comment]
> I will fix pep8 and retriever tests once #2568 is merged :)

it has been merged...

File: pytorch_lightning/core/step_result.py [hunk 1]
```diff
@@ -116,37 +120,22 @@ def log(
...
     def __set_meta(
-            self,
-            name: str,
-            value: Any,
-            prog_bar: bool,
-            logger: bool,
-            on_step: bool,
-            on_epoch: bool,
-            reduce_fx: Callable,
+        self, name: str, value: Any, prog_bar: bool, logger: bool, on_step: bool, on_epoch: bool, reduce_fx: Callable,
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:123]: lets keep per line

File: pytorch_lightning/core/step_result.py [hunk 2]
```diff
@@ -23,30 +24,26 @@ def __init__(
         if hiddens is not None:
             self.hiddens = hiddens
         if minimize is not None:
-            err = 'Minimize can only be used in training_step, training_step_end, training_epoch_end'
-            self._assert_grad_tensor_metric('minimize', minimize, err)
+            err = "Minimize can only be used in training_step, training_step_end, training_epoch_end"
```
Comment (Borda) [inline pytorch_lightning/core/step_result.py:27]: all these replacements make the PR messy - unclear what is the change/PR about

**Comment by justusschock** [comment]
Okay, I'll revert those changes. That was automatically formatted, since there were some pep8 issues :D

**Comment by Borda** [comment]
> Okay, I'll revert those changes. That was automatically formatted, since there were some pep8 issues :D

I think that pep8, is not soo strict about " and ' and as there is no wider consensus which shall be used I would drop this format changes... but you are free to re-open the question and et a conclusion :]

**Comment by xiadingZ** [comment]
Will this `sync_ddp_op` in result conflict with `reduce_op` in TensorMetric?
If I set 'reduce_op' of TensorMetric as `SUM`,  then set `sync_ddp_op` in Result as 'mean', what will happen?

**Comment by justusschock** [comment]
@xiadingZ When you reduce your metric with SUM, you will get what you expected (a sum across all processes in DDP) on every node. When you now sync it again with `Result` it will now give you the mean across all projects (which should not change anything in this case, since all values should be identical now, however if you use an additional SUM here, it changes something). 

The general way to go here is: Whenever something has built-in support for DDP Sync, use this one, since we may have customised it (currently we are working on providing a more useful way to sync certain metrics). If not, use the sync from results.

**Pull request merged**